upstash_redis/client.py

```diff
@@ -117,11 +116,125 @@ def execute(self, command: List) -> RESTResultT:
             command=command,
         )
 
-        main_command = command[0]
-        if len(command) > 1 and main_command == "SCRIPT":
-            main_command = f"{main_command} {command[1]}"
+        return cast_response(command, res)
+
+    def pipeline(self) -> "Pipeline":
+        """
+        Create a pipeline to send commands in batches
+        """
+        return Pipeline(
+            url=self._url,
+            token=self._token,
+            rest_encoding=self._rest_encoding,
+            rest_retries=self._rest_retries,
+            rest_retry_interval=self._rest_retry_interval,
+            allow_telemetry=self._allow_telemetry,
+            headers=self._headers,
+            session=self._session,
+            multi_exec="pipeline"
+        )
+
+    def multi(self) -> "Pipeline":
+        """
+        Create a pipeline to send commands in batches as a transaction
+        """
+        return Pipeline(
+            url=self._url,
+            token=self._token,
+            rest_encoding=self._rest_encoding,
+            rest_retries=self._rest_retries,
+            rest_retry_interval=self._rest_retry_interval,
+            allow_telemetry=self._allow_telemetry,
+            headers=self._headers,
+            session=self._session,
+            multi_exec="multi-exec"
+        )
+
+
+class Pipeline(PipelineCommands):
+
+    def __init__(
+        self,
+        url: str,
+        token: str,
+        rest_encoding: Optional[Literal["base64"]] = "base64",
+        rest_retries: int = 1,
+        rest_retry_interval: float = 3,  # Seconds.
+        allow_telemetry: bool = True,
+        headers: Optional[Dict[str, str]] = None,
+        session: Optional[Session] = None,
+        multi_exec: Literal["multi-exec", "pipeline"] = "pipeline"
+    ):
+        """
+        Creates a new blocking Redis client.
+
+        :param url: UPSTASH_REDIS_REST_URL in the console
+        :param token: UPSTASH_REDIS_REST_TOKEN in the console
+        :param rest_encoding: the encoding that can be used by the REST API to parse the response before sending it
+        :param rest_retries: how many times an HTTP request will be retried if it fails
+        :param rest_retry_interval: how many seconds will be waited between each retry
+        :param allow_telemetry: whether anonymous telemetry can be collected
+        :param headers: request headers
+        :param session: A Requests session
+        :param miltiexec: Whether multi execution (transaction) or pipelining is to be used
+        """
+
+        self._url = url
+        self._token = token
+
+        self._allow_telemetry = allow_telemetry
+
+        self._rest_encoding: Optional[Literal["base64"]] = rest_encoding
+        self._rest_retries = rest_retries
+        self._rest_retry_interval = rest_retry_interval
+
+        self._headers = headers or make_headers(token, rest_encoding, allow_telemetry)
+        self._session = session or Session()
```

Comment:
I think we should stick with using the session of the client, because it means that we might re-use some of the already initialized connections that are being kept alive in the connection pool. Creating a new session should also work, but it might cause higher latency. BTW, regarding not sharing the context in the async mode, did you faced with an issue or was it just a gut feeling?

Also, we can reuse the headers from the client, no need to recreate those, as they will always be the same